// @nuxtjs/google-fonts/dist/module.mjs
import { resolve } from 'pathe';
import { useLogger, defineNuxtModule, isNuxt2, resolvePath } from '@nuxt/kit';
import { isValidURL, parse, constructURL, merge, download } from 'google-fonts-helper';

const name = "@nuxtjs/google-fonts";
const version = "3.0.0-0";

const logger = useLogger("nuxt:google-fonts");
const module = defineNuxtModule({
  meta: {
    name,
    version,
    configKey: "googleFonts"
  },
  defaults: {
    families: {},
    display: void 0,
    subsets: [],
    text: void 0,
    prefetch: true,
    preconnect: true,
    preload: false,
    useStylesheet: false,
    download: true,
    base64: false,
    inject: true,
    overwriting: false,
    outputDir: "node_modules/.cache/nuxt-google-fonts",
    stylePath: "css/nuxt-google-fonts.css",
    fontsDir: "fonts",
    fontsPath: "../fonts"
  },
  async setup(options, nuxt) {
    if (options.display === void 0 && !options.preload) {
      options.display = "swap";
    }
    const fontsParsed = [];
    if (isNuxt2()) {
      nuxt.options.head = nuxt.options.head || {};
      nuxt.options.head.link = nuxt.options.head.link || [];
      fontsParsed.push(...nuxt.options.head.link.filter((link) => isValidURL(link.href)).map((link) => parse(link.href)));
    } else {
      nuxt.options.app.head.link = nuxt.options.app.head.link || [];
      fontsParsed.push(...nuxt.options.app.head.link.filter((link) => isValidURL(link.href)).map((link) => parse(link.href)));
    }
    const url = constructURL(merge(options, ...fontsParsed));
    if (!url) {
      logger.warn("No provided fonts.");
      return;
    }
    if (isNuxt2()) {
      nuxt.options.head = nuxt.options.head || {};
      nuxt.options.head.link = nuxt.options.head.link || [];
      nuxt.options.head.link = nuxt.options.head.link.filter((link) => !isValidURL(link.href));
    } else {
      nuxt.options.app.head.link = nuxt.options.app.head.link || [];
      nuxt.options.app.head.link = nuxt.options.app.head.link.filter((link) => !isValidURL(link.href));
    }
    if (options.download) {
      const outputDir = await resolvePath(options.outputDir);
      try {
        const downloader = download(url, {
          base64: options.base64,
          overwriting: options.overwriting,
          outputDir,
          stylePath: options.stylePath,
          fontsDir: options.fontsDir,
          fontsPath: options.fontsPath
        });
        await downloader.execute();
        if (options.inject) {
          nuxt.options.css.push(resolve(outputDir, options.stylePath));
        }
        nuxt.options.nitro = nuxt.options.nitro || {};
        nuxt.options.nitro.publicAssets = nuxt.options.nitro.publicAssets || [];
        nuxt.options.nitro.publicAssets.push({ dir: outputDir });
      } catch (e) {
        logger.error(e);
      }
      return;
    }
    if (isNuxt2()) {
      nuxt.options.head = nuxt.options.head || {};
      nuxt.options.head.link = nuxt.options.head.link || [];
      nuxt.options.head.script = nuxt.options.head.script || [];
      nuxt.options.head.noscript = nuxt.options.head.noscript || [];
      nuxt.options.head.__dangerouslyDisableSanitizersByTagID = nuxt.options.head.__dangerouslyDisableSanitizersByTagID || {};
      if (options.prefetch) {
        nuxt.options.head.link.push({
          hid: "gf-prefetch",
          rel: "dns-prefetch",
          href: "https://fonts.gstatic.com/"
        });
      }
      if (options.preconnect) {
        nuxt.options.head.link.push({
          hid: "gf-preconnect",
          rel: "preconnect",
          href: "https://fonts.gstatic.com/",
          crossorigin: "anonymous"
        }, {
          hid: "gf-origin-preconnect",
          rel: "preconnect",
          href: "https://fonts.googleapis.com/"
        });
      }
      if (options.preload) {
        nuxt.options.head.link.push({
          hid: "gf-preload",
          rel: "preload",
          as: "style",
          href: url
        });
      }
      if (options.useStylesheet) {
        nuxt.options.head.link.push({
          hid: "gf-style",
          rel: "stylesheet",
          href: url
        });
        return;
      }
      nuxt.options.head.script.push({
        hid: "gf-script",
        innerHTML: `(function(){var l=document.createElement('link');l.rel="stylesheet";l.href="${url}";document.querySelector("head").appendChild(l);})();`
      });
      nuxt.options.head.noscript.push({
        hid: "gf-noscript",
        innerHTML: `<link rel="stylesheet" href="${url}">`
      });
      nuxt.options.head.__dangerouslyDisableSanitizersByTagID["gf-script"] = ["innerHTML"];
      nuxt.options.head.__dangerouslyDisableSanitizersByTagID["gf-noscript"] = ["innerHTML"];
      return;
    }
    nuxt.options.app.head.link = nuxt.options.app.head.link || [];
    nuxt.options.app.head.script = nuxt.options.app.head.script || [];
    if (options.prefetch) {
      nuxt.options.app.head.link.push({
        rel: "dns-prefetch",
        href: "https://fonts.gstatic.com/"
      });
    }
    if (options.preconnect) {
      nuxt.options.app.head.link.push({
        rel: "preconnect",
        href: "https://fonts.gstatic.com/",
        crossorigin: "anonymous"
      }, {
        rel: "preconnect",
        href: "https://fonts.googleapis.com/"
      });
    }
    if (options.preload) {
      nuxt.options.app.head.link.push({
        rel: "preload",
        as: "style",
        href: url
      });
    }
    if (options.useStylesheet) {
      nuxt.options.app.head.link.push({
        rel: "stylesheet",
        href: url
      });
      return;
    }
    nuxt.options.app.head.script.unshift({
      "data-hid": "gf-script",
      children: `(function(){
        var h=document.querySelector("head");
        var m=h.querySelector('meta[name="head:count"]');
        if(m){m.setAttribute('content',Number(m.getAttribute('content'))+1);}
        else{m=document.createElement('meta');m.setAttribute('name','head:count');m.setAttribute('content','1');h.append(m);}
        var l=document.createElement('link');l.rel='stylesheet';l.href='${url}';h.appendChild(l);
      })();`
    });
  }
});

export { module as default };
